internal/api/installation.go

```diff
@@ -325,21 +325,57 @@ func handleJoinGroup(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusOK)
 }
 
-// handleLeaveGroup responds to DELETE /api/installation/{installation}/group, leaving any existing group.
+// handleLeaveGroup responds to DELETE /api/installation/{installation}/group,
+// leaving any existing group.
 func handleLeaveGroup(c *Context, w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	installationID := vars["installation"]
 	c.Logger = c.Logger.WithField("installation", installationID)
 
+	retainConfig, err := parseBool(r.URL, "retain_config", true)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to parse retain_config setting")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
 	installation, status, unlockOnce := lockInstallation(c, installationID)
 	if status != 0 {
 		w.WriteHeader(status)
 		return
 	}
 	defer unlockOnce()
 
+	// TODO: does it make sense to enforce normal update-requested valid states?
+	// Should there be more or less valid states? Review this when necessary.
+	newState := model.InstallationStateUpdateRequested
+	if !installation.ValidTransitionState(newState) {
+		c.Logger.Warnf("unable to leave group while installation is in state %s", installation.State)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
 	if installation.GroupID != nil {
+		installation.State = newState
 		installation.GroupID = nil
+		installation.GroupSequence = nil
+
+		if retainConfig {
+			// The installation is leaving the group, but the config is being set
+			// to the group-merged version used while it was in the group. To do
+			// so, we will get a merged copy of the installation out and will
+			// manually update the necessary values.
+			mergedInstallation, err := c.Store.GetInstallation(installationID, true, false)
+			if err != nil {
+				c.Logger.WithError(err).Error("failed to get group-merged installation")
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+
+			installation.Version = mergedInstallation.Version
```

Comment:
Refresh my memory here, but in normal operation, is the merge config written to the Installation struct, overwriting whatever configuration may have been there before the Installation joined the group? Or is it safe to assume that until this line, `installation.Version` will refer to the version the Installation had before joining the group, and that data is only actually overwritten here?

Reply:
The answer is: it totally depends on how you ask for the installation.

https://github.com/mattermost/mattermost-cloud/blob/c1621a2453a697d1d830be07f2dd7bf90f35a90e/internal/store/installation.go#L61

The function for getting installations will do whatever you tell it to so you can ask for what you want.

Reply:
OK cool so we don't ever overwrite the values in the Installation fields themselves, new values are always stored in the Group and we just check the overrides values (I do remember this from your previous PRs). Perfect.

Reply:
Yeah, to be super clear. If you update the installation directly, the installation DB object is updated. For group updates, it's the group object that is updated. Then, when acting on installations it's always the merged set of what both are at that moment in time.